homeassistant/components/rachio.py

```diff
@@ -0,0 +1,282 @@
+
+CONF_CUSTOM_URL = 'hass_url_override'
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.Schema({
+        vol.Required(CONF_API_KEY): cv.string,
+        vol.Optional(CONF_CUSTOM_URL): cv.string,
+    })
+}, extra=vol.ALLOW_EXTRA)
+
+# Keys used in the API JSON
+KEY_DEVICE_ID = 'deviceId'
+KEY_DEVICES = 'devices'
+KEY_ENABLED = 'enabled'
+KEY_EXTERNAL_ID = 'externalId'
+KEY_ID = 'id'
+KEY_NAME = 'name'
+KEY_ON = 'on'
+KEY_STATUS = 'status'
+KEY_SUBTYPE = 'subType'
+KEY_SUMMARY = 'summary'
+KEY_TYPE = 'type'
+KEY_URL = 'url'
+KEY_USERNAME = 'username'
+KEY_ZONE_ID = 'zoneId'
+KEY_ZONE_NUMBER = 'zoneNumber'
+KEY_ZONES = 'zones'
+
+STATUS_ONLINE = 'ONLINE'
+STATUS_OFFLINE = 'OFFLINE'
+
+# Device webhook values
+TYPE_CONTROLLER_STATUS = 'DEVICE_STATUS'
+SUBTYPE_OFFLINE = 'OFFLINE'
+SUBTYPE_ONLINE = 'ONLINE'
+SUBTYPE_OFFLINE_NOTIFICATION = 'OFFLINE_NOTIFICATION'
+SUBTYPE_COLD_REBOOT = 'COLD_REBOOT'
+SUBTYPE_SLEEP_MODE_ON = 'SLEEP_MODE_ON'
+SUBTYPE_SLEEP_MODE_OFF = 'SLEEP_MODE_OFF'
+SUBTYPE_BROWNOUT_VALVE = 'BROWNOUT_VALVE'
+SUBTYPE_RAIN_SENSOR_DETECTION_ON = 'RAIN_SENSOR_DETECTION_ON'
+SUBTYPE_RAIN_SENSOR_DETECTION_OFF = 'RAIN_SENSOR_DETECTION_OFF'
+SUBTYPE_RAIN_DELAY_ON = 'RAIN_DELAY_ON'
+SUBTYPE_RAIN_DELAY_OFF = 'RAIN_DELAY_OFF'
+
+# Schedule webhook values
+TYPE_SCHEDULE_STATUS = 'SCHEDULE_STATUS'
+SUBTYPE_SCHEDULE_STARTED = 'SCHEDULE_STARTED'
+SUBTYPE_SCHEDULE_STOPPED = 'SCHEDULE_STOPPED'
+SUBTYPE_SCHEDULE_COMPLETED = 'SCHEDULE_COMPLETED'
+SUBTYPE_WEATHER_NO_SKIP = 'WEATHER_INTELLIGENCE_NO_SKIP'
+SUBTYPE_WEATHER_SKIP = 'WEATHER_INTELLIGENCE_SKIP'
+SUBTYPE_WEATHER_CLIMATE_SKIP = 'WEATHER_INTELLIGENCE_CLIMATE_SKIP'
+SUBTYPE_WEATHER_FREEZE = 'WEATHER_INTELLIGENCE_FREEZE'
+
+# Zone webhook values
+TYPE_ZONE_STATUS = 'ZONE_STATUS'
+SUBTYPE_ZONE_STARTED = 'ZONE_STARTED'
+SUBTYPE_ZONE_STOPPED = 'ZONE_STOPPED'
+SUBTYPE_ZONE_COMPLETED = 'ZONE_COMPLETED'
+SUBTYPE_ZONE_CYCLING = 'ZONE_CYCLING'
+SUBTYPE_ZONE_CYCLING_COMPLETED = 'ZONE_CYCLING_COMPLETED'
+
+# Webhook callbacks
+LISTEN_EVENT_TYPES = ['DEVICE_STATUS_EVENT', 'ZONE_STATUS_EVENT']
+WEBHOOK_CONST_ID = 'homeassistant.rachio'
+WEBHOOK_PATH = URL_API + DOMAIN
+EVENT_RACHIO_UPDATE = DOMAIN + '_update'
+EVENT_RACHIO_CONTROLLER_UPDATE = EVENT_RACHIO_UPDATE + '_controller'
+EVENT_RACHIO_ZONE_UPDATE = EVENT_RACHIO_UPDATE + '_zone'
+EVENT_RACHIO_SCHEDULE_UPDATE = EVENT_RACHIO_UPDATE + '_schedule'
+
+
+def setup(hass, config) -> bool:
+    """Set up the Rachio component."""
+    from rachiopy import Rachio
+
+    # Listen for incoming webhook connections
+    hass.http.register_view(RachioWebhookView())
+
+    # Configure API
+    api_key = config[DOMAIN].get(CONF_API_KEY)
+    rachio = Rachio(api_key)
+
+    # Get the URL of this server
+    custom_url = config[DOMAIN].get(CONF_CUSTOM_URL)
+    hass_url = hass.config.api.base_url if custom_url is None else custom_url
+    webhook_auth = '?api_password={}'.format(hass.config.api.api_password)
+    rachio.webhook_url = hass_url + WEBHOOK_PATH + webhook_auth
```

Comment:
How is this used?

Reply:
I'm not sure exactly what "this" is (and I'm not even writing JS...)

1. The custom URL can be set in the config.
1. If set, it overrides the base_url. If not, it stores the base_url.
1. (Now replaced with `generate_secret`.) This is passed to the Rachio API and then provided with incoming webhooks to authenticate the connections.
1. This assembles the hostname and path. The combine URL is then given to the Rachio API to call when events occur.